# 1
#### From Videos

In [4]:
import os
import pandas as pd
import numpy as np


In [5]:
df =pd.read_csv(r'D:\\Mazen Essam\\Machine Learning course\\ENG Mohamed Documentary\\products.csv',index_col=False) #read from a csv 

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64  
 4   prices         49693 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [7]:
df.head() #show first 5 rows in the dataframe


product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [8]:
df.tail() #show last 5 rows in the dataframe

product_id                                product_name  aisle_id  \
49688       49684   Vodka, Triple Distilled, Twist of Vanilla       124   
49689       49685          En Croute Roast Hazelnut Cranberry        42   
49690       49686                            Artisan Baguette       112   
49691       49687  Smartblend Healthy Metabolism Dry Cat Food        41   
49692       49688                      Fresh Foaming Cleanser        73   

       department_id  prices  
49688              5     5.3  
49689              1     3.1  
49690              3     7.8  
49691              8     4.7  
49692             11    13.5

In [9]:
df.dtypes # show datatypes of the dataframe

product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object

In [10]:
df.columns #show columns of the dataframe

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices'], dtype='object')

In [11]:
df.describe() # get some details about the dataframe

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [12]:
df.drop(columns=['prices']) #drop a column

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  
0            61             19  
1           104             13  
2            94              7  
3            38              1  
4             5             13  
...         ...            ...  
49688       124              5  
49689        42              1  
49690       112              3  
49691        41              8  
49692        73             11  

[49693 rows x 4 columns]

In [13]:
df['product_id'].value_counts(dropna=False) # show how many a data in a certain column repeated

product_id
6800     2
35491    2
18458    2
462      2
35306    2
        ..
16567    1
16568    1
16569    1
16570    1
49688    1
Name: count, Length: 49686, dtype: int64

In [14]:
df.rename(columns={"product_id":'ID'},inplace=True) # rename a column
df.head(10) # show first 10 of dataframe

ID                                       product_name  aisle_id  \
0   1                         Chocolate Sandwich Cookies        61   
1   2                                   All-Seasons Salt       104   
2   3               Robust Golden Unsweetened Oolong Tea        94   
3   4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4   5                          Green Chile Anytime Sauce         5   
5   6                                       Dry Nose Oil        11   
6   7                     Pure Coconut Water With Orange        98   
7   8                  Cut Russet Potatoes Steam N' Mash       116   
8   9                  Light Strawberry Blueberry Yogurt       120   
9  10     Sparkling Orange Juice & Prickly Pear Beverage       115   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3  
5             11     2.6  
6              7     4.4  
7              1     1.1  
8             16     7.0  
9              7     8.4

In [15]:
df['ID']=df['ID'].astype('float') #change data type

# 2

In [16]:
df_ords=pd.read_csv(r'D:\\Mazen Essam\\Machine Learning course\\ENG Mohamed Documentary\\orders.csv',index_col=False)

In [17]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


#### change any id to string as we don't need to make any operations on them 

In [21]:
df_ords['order_id']=df_ords['order_id'].astype('string') # change data type of id to be string
df_ords['user_id']=df_ords['user_id'].astype('string')
df_ords.dtypes

order_id                  string[python]
user_id                   string[python]
eval_set                          object
order_number                       int64
order_dow                          int64
order_hour_of_day                  int64
days_since_prior_order           float64
dtype: object

# 3

In [22]:
df_ords.rename(columns={"order_dow":'order_day'},inplace=True) # rename order_dow to order_day

In [24]:
df_ords.columns # show columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_day',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

# 4

In [25]:
HOURS=df_ords['order_hour_of_day'].value_counts(dropna=False) #get how many the data in order_hour_of_day repeated
HOURS.sort_index # sort data
print(f"Highst orders\' Hour in the day is {HOURS.keys()[0]}")

Highst orders' Hour in the day is 10


# 5

#### Read csv file 

In [55]:

df_prod=pd.read_csv(r'D:\\Mazen Essam\\Machine Learning course\\ENG Mohamed Documentary\\products.csv',index_col=False)


#### Converting the departments in the dataframe to dictionary 

In [56]:
df_new=dict()
mx=max(df_prod["department_id"])
mn=min(df_prod["department_id"])

for i in range(mn,mx+1) :
    arr=[]
    for j in range(len(df_prod)):
        if ((df_prod.iloc[j,3]))==i :
            arr.append(df_prod.iloc[j,1])
    df_new.update({i:arr})         




#### Showing keys of the dict


In [57]:
df_new.keys()


dict_keys([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21])

#### The values of key 4

In [58]:
df_new[4]

['White Pearl Onions',
 'Organic Clementines',
 'European Cucumber',
 'European Style Spring Mix',
 'Yogurt Fruit Dip Sliced Apples',
 'Local Living Butter Lettuce',
 'Cauliflower Florettes',
 'Arugula Salad',
 'Organic Lemons',
 'Nectarines',
 'Living Organic Cilantro',
 'Herb Thyme Clamshell',
 'Cantaloupe',
 'Gold Potatoes',
 'Juicing Carrots',
 'Organic Whole Peeled Baby Carrots',
 'Vegetable Salad Kit, Sweet Kale',
 'Greener Selection Salad',
 'Clementine Bag',
 'Mache Blend',
 'Organic Yukon Gold Potato Bag',
 'Kabocha Squash',
 'Yellow Cherry Tomato',
 'Red Seedless Grapes Imported',
 'Garnet Sweet Potato (Yam)',
 'Organic Crunchy Slaw Americana Salad',
 'Celery Heart',
 'Romaine',
 'Organic Vegetable Tray',
 'Organic Spinach Quinoa Powermeal Bowl',
 'Grilled Asparagus',
 'ArtiHearts Fresh Grilled Artichokes',
 'Fingerling Potato',
 'Organic Fresh Basil',
 'Organic Broccolini',
 'Salad, Cobb',
 'Mini Watermelon',
 'Organic Black Seedless Grapes',
 'Grapes',
 'Organic Orange Bell

data of department 4 consists of vegetables

# 6
#### Subset department_id=14 (Breakfast)

In [59]:
df_BF=df_prod.query('department_id==14') # making a subset according to numbet of department_id
df_BF

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

# 7
#### subset department_id= 5 , 20 , 7 , 12 (alcohol, deli, beverages, and meat/seafood)

In [60]:
df_DN=df_prod.query('department_id==[5,20,7,12]') # making a subset according to numbet of department_id


#### Sort the subset according to department_id

In [61]:
df_DN.sort_values(by=['department_id']) #sort values of department_id

product_id                               product_name  aisle_id  \
49688       49684  Vodka, Triple Distilled, Twist of Vanilla       124   
36276       36272                       Kali Hart Pinot Noir        28   
10802       10802                  Diamond Junmai Ginjo Saké       134   
10702       10702                        California Red Wine        28   
36423       36419                                   Cashmere        28   
...           ...                                        ...       ...   
15767       15767                    Blackened Chicken Salad         1   
26384       26383                          Beef Cotto Salami        96   
40692       40688                      Spinach Artichoke Dip        67   
32902       32900                        Paneer Tikka Masala        13   
7520         7520                  Tempeh, Organic, Soy Cake        14   

       department_id  prices  
49688              5     5.3  
36276              5    11.6  
10802              5     9.5  
10702              5     3.0  
36423              5     9.5  
...              ...     ...  
15767             20    11.1  
26384             20    11.2  
40692             20     1.7  
32902             20     2.3  
7520              20    10.0  

[7650 rows x 5 columns]

# 8
#### The last dataframe created (df_DN) consists of 7650 rows x 5 columns 

# 9
#### showing data of used_id 1

In [65]:
df_usr1=df_ords.query('user_id=="1"') # get data for user_data =1 and put it in new dataframe
df_usr1  # print the dataframe

order_id user_id eval_set  order_number  order_day  order_hour_of_day  \
0   2539329       1    prior             1          2                  8   
1   2398795       1    prior             2          3                  7   
2    473747       1    prior             3          3                 12   
3   2254736       1    prior             4          4                  7   
4    431534       1    prior             5          4                 15   
5   3367565       1    prior             6          2                  7   
6    550135       1    prior             7          1                  9   
7   3108588       1    prior             8          1                 14   
8   2295261       1    prior             9          1                 16   
9   2550362       1    prior            10          4                  8   
10  1187899       1    train            11          4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# 10 
#### Get most freq day and hour of User 1

In [76]:
df_usr1[["order_day","order_hour_of_day"]].mode() #mode() is used to get most freq data

order_day  order_hour_of_day
0        4.0                  7
1        NaN                  8

User 1 frequently order at day= 4 and hour= 7am